sound/soc/sof/intel/hda.c

```diff
@@ -251,10 +254,16 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
 	}
 
 	/* codec detection */
-	if (!bus->codec_mask)
+	if (!bus->codec_mask) {
 		dev_info(bus->dev, "no hda codecs found!\n");
-	else
-		dev_info(bus->dev, "hda codecs found, mask %lx!\n", bus->codec_mask);
+	} else {
+		mach = snd_soc_acpi_intel_hda_machines;
+		mach->sof_fw_filename = pdata->desc->nocodec_fw_filename;
```

Comment:
this looks wrong. You are overriding the machine already selected based on ACPI ID. this should only happen if there was no machine found earlier.

Reply:
We don't set  .sof_fw_filename in hda_machines, I think that's why we need set it above.
Please see below:
```
struct snd_soc_acpi_mach snd_soc_acpi_intel_hda_machines[] = {
	{
		/* .id is not used in this file */
		.drv_name = "skl_hda_dsp_generic",

		/* .fw_filename is dynamically set in skylake driver */

		/* .sof_fw_filename is dynamically set in sof/intel driver */

		.sof_tplg_filename = "intel/sof-hda-generic.tplg",
...
	},
	{},
};
```

We actually use same FW filename for different machines with same chip, e.g. "sof-apl.ri" for all I2S, HDA, nocodec machines.

```
struct sof_dev_desc {
...
	/* defaults for no codec mode */
	const char *nocodec_fw_filename; 
	const char *nocodec_tplg_filename;
...
};
```

So, how about rename struct sof_dev_desc.nocodec_fw_filename to "sof_dev_desc.default_fw_filename"?

```
struct sof_dev_desc {
...
	/* default binary names */
	const char *default_fw_filename; 
	const char *nocodec_tplg_filename;
...
};
```

Reply:
Another option is get sof_fw_filename from other I2S machine, but this is still not natural
````
mach = snd_soc_acpi_intel_hda_machines;
mach->sof_fw_filename = pdata->desc->machines[0].sof_fw_filename;
pdata->machine = mach;
````

Reply:
No. The feedback was on the No. The feedback was mach = snd_soc_acpi_intel_hda_machines;
You essentially remove all capability of using I2S codecs if there is HDMI detected. 
Please test on Up2 w/ Hifiberry.

Reply:
> No. The feedback was on the No. The feedback was mach = snd_soc_acpi_intel_hda_machines;
> You essentially remove all capability of using I2S codecs if there is HDMI detected.
> Please test on Up2 w/ Hifiberry.

ah, right, good catch. @RanderWang we should do this codec_mask checking and pdata->machine updating only at "pdata->machine == NULL" case.